controller/controller_server/utils.py

```diff
@@ -8,43 +10,164 @@
+        parameters_by_system = _parse_raw_input(raw_input=raw_parameters_by_system)
+        system_parameters = parameters_by_system.get(system_id, {})
+    default_pool = parameters.get(config.PARAMETERS_POOL)
+    default_space_efficiency = parameters.get(config.PARAMETERS_SPACE_EFFICIENCY)
+    default_prefix = parameters.get(prefix_param_name)
+    return ObjectParameters(
+        pool=system_parameters.get(config.PARAMETERS_POOL, default_pool),
+        space_efficiency=system_parameters.get(config.PARAMETERS_SPACE_EFFICIENCY, default_space_efficiency),
+        prefix=system_parameters.get(prefix_param_name, default_prefix))
+
+
+def get_volume_id(new_volume, system_id):
+    return _get_object_id(new_volume, system_id)
 
 
 def get_snapshot_id(new_snapshot):
     return _get_object_id(new_snapshot)
 
 
-def _get_object_id(obj):
+def _get_object_id(obj, system_id=None):
+    if system_id:
+        return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, system_id, obj.id))
     return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, obj.id))
 
 
-def validate_secret(secret):
-    logger.debug("validating secrets")
-    if secret:
-        if not (config.SECRET_USERNAME_PARAMETER in secret and
-                config.SECRET_PASSWORD_PARAMETER in secret and
-                config.SECRET_ARRAY_PARAMETER in secret):
-            raise ValidationException(messages.invalid_secret_message)
+def _is_imported_string_valid(imported_string):
+    return imported_string and re.match(config.SECRET_VALIDATION_REGEX, imported_string)
+
+
+def _validate_system_id(system_id):
+    if not _is_imported_string_valid(system_id):
+        raise ValidationException(messages.invalid_system_id_value_message.format(system_id))
+    if len(system_id) > config.SECRET_SYSTEM_ID_MAX_LENGTH:
+        raise ValidationException(
+            messages.parameter_length_is_too_long.format("system id", system_id, config.SECRET_SYSTEM_ID_MAX_LENGTH))
+
+
+def _validate_secrets(secrets):
+    if not (config.SECRET_USERNAME_PARAMETER in secrets and
+            config.SECRET_PASSWORD_PARAMETER in secrets and
+            config.SECRET_ARRAY_PARAMETER in secrets):
+        raise ValidationException(messages.invalid_secrets_message)
+
 
+def _validate_topologies(topologies):
+    if topologies:
+        for topology in topologies:
+            if not topology:
+                raise ValidationException(messages.invalid_secrets_message)
     else:
-        raise ValidationException(messages.secret_missing_message)
+        raise ValidationException(messages.invalid_secrets_message)
 
-    logger.debug("secret validation finished")
+
+def _validate_secrets_config(secrets_config):
+    for system_id, system_info in secrets_config.items():
+        if system_id and system_info:
+            _validate_topologies(system_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER))
+            _validate_system_id(system_id)
+            _validate_secrets(system_info)
```

Comment:
I would order them by order of appearance in the example, i.e. system id, system info "secrets", then topologies

Reply:
yes, but if no topologies, we don't care about the rest..

Reply:
isn't this true for each of them?

Reply:
no we can be fine without system_id and even if the secrets are valid it doesn't matter if we don't have topologies..

Reply:
1. if I rearrange the second clause into `if topologies are valid it doesn't matter if we don't have the secrets` - isn't it also true?
2. 

> we can be fine without system_id

the scope of this code is under `if system_id`, and even if topologies are valid it doesn't matter if the system_id is invalid

Reply:
ok, it in the `appearance in the example` order now